### DISCUSSION TITLE
Allow retiming messages in User Scripts

### DISCUSSION THREAD

**Discussion by jameskuszmaul-brt** [foxglove/community#594] [Q&A]
This came up in the community slack, but a not-uncommon situation to be in is to have a sensor message of some sort that includes a capture time and the sensor data. This capture time is generally different than the message publish time or the message log time, but it may be desirable to have the data in the sensor message visualized using the capture time rather than the actual message time.

To support this, I think the most straightforward option from the user's perspective would be to allow the `script` method in the User Script to return some special type that includes a timestamp field in addition to the new data (akin to how the `Input<>` has timestamps and topic information).

This would also support a use-case where sensor readings are batched into a single message (e.g., in https://github.com/frc971/971-Robot-Code/blob/master/frc971/wpilib/imu_batch.fbs we batch up 2 kHz IMU readings which we only actually process at ~200 Hz), with a message of a format akin to the linked flatbuffer schema:

```
table IMUValues {
  // Gyro readings in radians/second.
  // Positive is clockwise looking at the connector.
  gyro_x:float (id: 0);
  // Positive is clockwise looking at the right side (from the connector).
  gyro_y:float (id: 1);
  // Positive is counterclockwise looking at the top.
  gyro_z:float (id: 2);

  // Accelerometer readings in Gs.
  // Positive is up.
  accelerometer_x:float (id: 3);
  // Positive is away from the right side (from the connector).
  accelerometer_y:float (id: 4);
  // Positive is away from the connector.
  accelerometer_z:float (id: 5);

  // Magnetometer readings in gauss.
  // Positive is up.
  magnetometer_x:float (id: 6);
  // Positive is away from the right side (from the connector).
  magnetometer_y:float (id: 7);
  // Positive is away from the connector.
  magnetometer_z:float (id: 8);

  // Barometer readings in pascals.
  barometer:float (id: 9);

  // Temperature readings in degrees Celsius.
  temperature:float (id: 10);

  // FPGA timestamp when the values were captured.
  fpga_timestamp:double (id: 11);
  // CLOCK_MONOTONIC time in nanoseconds when the values were captured,
  // converted from fpga_timestamp.
  monotonic_timestamp_ns:long (id: 12);
}

table IMUValuesBatch {
  readings:[IMUValues] (id: 0);
}
```

**Comment by james-rms** [comment]
First impression, hot take: this would require a lot of changes in Foxglove Studio, and might not be possible without rearchitecting the player completely. Message buffering and preloading all depend heavily on the log time of the messages being read, and updating that timestamp within the message pipeline would break some key assumptions.

To serve this use-case, I think it'd be cleaner to rewrite the message log times in a script using the MCAP API before playing it back in Studio. This way Studio can correctly use the index of the rewritten MCAP to find messages by their (new, synthetic) log times.

**Reply by uforic** [reply]
"rewrite the message log times in a script using the MCAP API" - we actually went with this approach, using the Python MCAP SDK (just easier to do reflection than the go one), and passing in a map like: "topic/name" -> "inner.timestamp.field". I could image this being part of the mcap CLI, as just one of these common things people are doing with mcap all the time (like filtering), but could also see it standing alone. A little hard to fully generalize it, you run into things like "ah well this timestamp field is an int, whereas this one is a google.Timestamp seconds and nanos" one.
